src/libstd/fs.rs

```diff
@@ -1069,13 +1069,43 @@ pub fn remove_dir_all<P: AsRef<Path>>(path: P) -> io::Result<()> {
 }
 
 fn _remove_dir_all(path: &Path) -> io::Result<()> {
+    let filetype = try!(symlink_metadata(path)).file_type();
+    if filetype.is_symlink() {
+        if cfg!(windows) {
+            // On Windows symlinks to files and directories removed differently
+            // we should remove only directories here and have error on file
+            remove_dir(path)
+        } else {
+            // On unix symlinks are always files
+            remove_file(path)
+        }
+    } else {
+        remove_dir_all_recursive(path)
+    }
+}
+fn remove_dir_all_recursive(path: &Path) -> io::Result<()> {
     for child in try!(read_dir(path)) {
-        let child = try!(child).path();
-        let stat = try!(symlink_metadata(&*child));
-        if stat.is_dir() {
-            try!(remove_dir_all(&*child));
+        let child = try!(child);
+        let child_path = child.path();
+        let child_type = try!(child.file_type());
+        if child_type.is_dir() {
+            try!(remove_dir_all_recursive(&*child_path));
         } else {
-            try!(remove_file(&*child));
+            if cfg!(windows) {
+                if child_type.is_symlink() {
+                    let target_type = try!(metadata(&*child_path)).file_type();
```

Comment:
I tested this out a bit, and oddly enough I'm not sure we can use this function to determine this. Using `metadata` will resolve all intermediate symlinks, showing the file type of the actual destination file. This ends up having what I think are two problems:

1. If the symlink is broken (e.g. the destination is gone), I don't think this will work.
2. Removal of the symlink depends on what kind of symlink was created, not the destination itself.

I found that I could create a symlink to a file with `symlink_dir`, but I could only remove the symlink with `remove_dir`. I think that we may have enough information in the `DirEntry` itself to figure this out (via a private API for now).